### ISSUE TITLE
Support native approvals

### ISSUE THREAD

**Issue by prayerslayer** [zalando/zappr#442]
So GitHub just got [native approvals](https://github.com/blog/2256-a-whole-new-github-universe-announcing-new-tools-forums-and-features), which is great! We should support them once they are available in web hooks (currently they aren't, as usual).

**Comment by prayerslayer** [comment]
Update on this:

Reviews are now supported in the web hooks :tada: However in order to replace our existing workflow we also need "require reviews before merging" to be in the Branches API, which is as of now not the case.

**Comment by lukasniemeier-zalando** [comment]
But can we support "zappr approvals" inside the native reviews?

![image](https://cloud.githubusercontent.com/assets/10497901/20299472/5f711106-ab1b-11e6-85da-a95e87debfec.png)

**Comment by ePaul** [comment]
I think Zappr could just accept a native approval as an alternative to a comment with :+1:?

**Comment by prayerslayer** [comment]
> @lukasniemeier-zalando: But can we support "zappr approvals" inside the native reviews?

![https://media.giphy.com/media/EldfH1VJdbrwY/giphy.gif](https://media.giphy.com/media/EldfH1VJdbrwY/giphy.gif)

Yes.

**Comment by prayerslayer** [comment]
So to give an update on this. Technically it's now possible to integrate native approvals into Zappr. However preliminary testing revealed some oddities about GH approvals, like this one:

![image](https://cloud.githubusercontent.com/assets/878512/22586814/23c04fa0-e9ff-11e6-8282-1402c96a8542.png)

Integrating with GH approvals is useless as long as this can happen.

**Comment by j0k3r** [comment]
This is a really bad behavior but it seems that they are _aware_ of it.
https://twitter.com/GitHubHelp/status/806641222630703104

**Comment by prayerslayer** [comment]
Ahh, thanks for asking ❤️  So probably it will "just work" one day without prior announcement :grin:

**Comment by ePaul** [comment]
@prayerslayer Not sure how the native approvals look in the API, but if you can get their date (or their relative order in comparison to the pushes), we should still be able to handle a native approval just as a ":+1:" comment, and thus use the existing approval verification logic?

**Comment by prayerslayer** [comment]
No, [apparently](https://developer.github.com/v3/pulls/reviews/#get-a-single-review) there's no timestamp.

Also we would run into the situation where there's a GH approval saying everything's fine and Zappr saying that a review is missing... Not the best UX.

**Comment by ePaul** [comment]
Hmm, no timestamp, but a commit id. So we could compare that commit id with the pull request's head sha to see whether the approved commit was the latest one.

I agree that the UX is not the best, but the current state (the approvals are not considered at all) is even worse. (Possibly Zappr's message could be changed to say something like "There were commits after the last approval, please approve again" in this case.)

**Comment by lukasniemeier-zalando** [comment]
Maybe I am missing the point, but coming back to my comment https://github.com/zalando/zappr/issues/442#issuecomment-260586264 - isn't the "review" itself a "comment" (with a timestamp and all the bells)? So if I put a :+1: as my "review summary", can zappr count this as an approval correctly?

**Comment by prayerslayer** [comment]
> isn't the "review" itself a "comment" (with a timestamp and all the bells)

It's not. It's a "review" which is different from a "review comment" which is different from an "issue comment" (=> which is what Zappr relies on).

> So we could compare that commit id with the pull request's head sha to see whether the approved commit was the latest one.

Yeah I guess that would work. But I feel it's effort to fix something that shouldn't work that way in the first place. I'd rather sit it out, wait for GH to fix their review process and then do a clean transition, where a native approval "just works".

**Comment by mvalkon** [comment]
@prayerslayer @ePaul @lukasniemeier-zalando seems like GH has fixed this problem at least partially https://github.com/isaacs/github/issues/783

If we can live with the preconditions to use this (master branch protection enabled with the options mentioned in the issue) then I think we could enable this. Any thoughts?

**Comment by ruudk** [comment]
Any news on this? I work on a lot of PR's for Skipper, and it's super annoying to have all my PR's always in orange state because zappr checks are not yet finalized.

**Comment by yousifalraheem** [comment]
> So to give an update on this. Technically it's now possible to integrate native approvals into Zappr. However preliminary testing revealed some oddities about GH approvals, like this one:
> 
> ![image](https://cloud.githubusercontent.com/assets/878512/22586814/23c04fa0-e9ff-11e6-8282-1402c96a8542.png)
> 
> Integrating with GH approvals is useless as long as this can happen.

Actually, Github allows you to dismiss PR reviews when new commits are pushed. You just have to configure it.

<img width="917" alt="Screenshot 2021-04-20 at 10 53 24" src="https://user-images.githubusercontent.com/10633734/115367696-b956c480-a1c6-11eb-983f-c3ba38355726.png">

**Comment by martinpallmann** [comment]
Man I'd love to have this.

**Comment by red-avtovo** [comment]
I'm afraid to ask if there is any updates on that issue
